### ISSUE TITLE
Discuss external linking and linking order

### ISSUE THREAD

**Issue by petrochenkov** [blas-lapack-rs/lapacke-sys#2]
So you get linking errors when using `lapacke-sys` without some other crate linking to `liblapacke`.

This was found when testing some linking order changes in `rustc` in https://github.com/rust-lang/rust/pull/102832#issuecomment-1279772306 (build failure - https://crater-reports.s3.amazonaws.com/pr-102832-1/try%230c94de7c1a8f4b82b92d562847c35f513f520c7a/gh/lopez86.rust-mlearn/log.txt).

The dependency on `liblapacke` should be declared using a [`#[link]`](https://doc.rust-lang.org/reference/items/external-blocks.html?highlight=link#the-link-attribute) attribute or a build script with [`cargo:rustc-link-lib`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-liblib).

**Comment by IvanUkhov** [comment]
This is correct. The crate is not doing any linkage. Please refer to the [architecture](https://blas-lapack-rs.github.io/architecture).

**Comment by petrochenkov** [comment]
That architecture worked largely by luck so far.
You can see from the build failure log that the crate that *uses* symbols (lapacke-sys) doesn't declare a dependency on a library that *defines* those symbols (liblapacke.(so,a)). In general, `rustc` cannot build a correct linker command line in such situation.

**Comment by IvanUkhov** [comment]
There might be some misunderstanding on my side. The linking happens when one chooses a source. In the case of the Netlib source, for instance, the corresponding instructions are given here:

https://github.com/blas-lapack-rs/netlib-src/blob/master/build.rs

Are you referring to this or something else?

**Comment by tari** [comment]
It looks like you've got the same problem as [llvm-sys](https://gitlab.com/taricorp/llvm-sys.rs/-/issues/41), where the set of linked libraries cannot be known until build-time. It seems like we'd need Cargo to support emitting an equivalent of `links` in the manifest from build scripts to make this work correctly.

**Comment by IvanUkhov** [comment]
Yes, this is accurate. Thank you for looking into this!

**Comment by petrochenkov** [comment]
I'll also look in more detail how this crate and llvm-sys setup things later this week.

> The linking happens when one chooses a source.

I guess the main source of misunderstanding may come from the fact that it's not enough to link to the necessary library *somewhere* on the command line, it needs to be linked *after* it's used due to the order-dependent way in which unix-style linkers work (and the order is inferred from "crate -> crate" and "crate -> native lib" dependencies in rustc using topological sorting).

**Comment by IvanUkhov** [comment]
Thank you. Let us add the following to the list:

- `blas-sys`,
- `cblas-sys`,
- `lapack-sys`, and
- `lapacke-sys`.
